<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [56]:
import pandas as pd
import psycopg2

In [58]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [59]:
# текст запроса
query_3_1 = f'''select
                count(id) cnt_vac
                from vacancies
'''

In [60]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

/tmp/ipykernel_11359/1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


cnt_vac
0    49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [61]:
# текст запроса
query_3_2 = f'''select
                count(id) cnt_emp
                from employers
'''

In [62]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

/tmp/ipykernel_11359/3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


cnt_emp
0    23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [63]:
# текст запроса
query_3_3 = f'''select
                count(id) cnt_areas
                from areas
'''

In [64]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

/tmp/ipykernel_11359/928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


cnt_areas
0       1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [65]:
# текст запроса
query_3_4 = f'''select
                count(id) cnt_ind
                from industries
'''

In [66]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

/tmp/ipykernel_11359/2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


cnt_ind
0      294

***

# Выводы по предварительному анализу данных
Предварительный анализ показал, что база данных содержит более 49197 вакансий в 294 сферах деятельности от 23501 работодателей из 1362 городов (регионов).


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [67]:
# текст запроса
query_4_1 = f'''select
                count(v.id) cnt,
                a.name area
                from vacancies v
                join areas a on v.area_id = a.id
                group by a.name
                order by cnt desc
'''

In [68]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

/tmp/ipykernel_11359/2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


cnt             area
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [69]:
# текст запроса
query_4_2 = f'''select
                count(v.id) cnt_vac_salary
                from vacancies v
                where v.salary_from is not null
                or v.salary_to is not null
'''

In [70]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

/tmp/ipykernel_11359/970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


cnt_vac_salary
0           24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [71]:
# текст запроса
query_4_3 = f'''select
                round(avg(v.salary_from)) salary_from,
                round(avg(v.salary_to)) salary_to
                from vacancies v
                
'''

In [72]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

/tmp/ipykernel_11359/2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


salary_from  salary_to
0      71065.0   110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [73]:
# текст запроса
query_4_4 = f'''select
                count(v.id) cnt,
                v.schedule,
                v.employment
                from vacancies v
                group by v.schedule, v.employment
                order by cnt desc
'''

In [74]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

/tmp/ipykernel_11359/2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


cnt          schedule           employment
0   35367       Полный день     Полная занятость
1    7802  Удаленная работа     Полная занятость
2    1593     Гибкий график     Полная занятость
3    1312  Удаленная работа  Частичная занятость
4     940    Сменный график     Полная занятость
5     569       Полный день           Стажировка
6     367    Вахтовый метод     Полная занятость
7     347       Полный день  Частичная занятость
8     312     Гибкий график  Частичная занятость
9     141       Полный день     Проектная работа
10    133  Удаленная работа     Проектная работа
11    116     Гибкий график           Стажировка
12    101    Сменный график  Частичная занятость
13     64  Удаленная работа           Стажировка
14     18     Гибкий график     Проектная работа
15     12    Сменный график           Стажировка
16      2    Вахтовый метод     Проектная работа
17      1    Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [75]:
# текст запроса
query_4_5 = f'''select
                v.experience
                from vacancies v
                group by v.experience
                order by count(v.id)
'''

In [76]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

/tmp/ipykernel_11359/2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

# Выводы по детальному анализу вакансий
Лидерами по количеству размещенных вакансий являются компании из крупных городов (Москва, Санкт-Петербург, Минск, Новосибирск, Алматы) (это логично). Как видим, база данных содержит вакансии не только российских, но и иностранных работодателей.
Заработная плата указана менее, чем в половине вакансий (24073). Средняя зарплатная вилка среди заполненных полей - от 71065 руб. до 110537 руб. в месяц. Большинство вакансий (35367) размещено для поиска сотрудников на полный рабочий день и полную занятость. На втором месте (7802) находятся вакансии, предлагающие удаленную работу и полную занятость. Большинство вакансий ищут кандидатов с опытом работы от 1 года до 3 лет. Требуемый опыт более 6 лет встречается в вакансиях реже всего. Видимо, такой опыт требуется на руководящих должностях, вакансии на которые не столь многочисленны. И, конечно, редко встречаются вакансии для кандидатов без опыта работы,т.к. работодатели заинтересованы в кандидатах, которые уже имеют минимальный опыт, и их не надо обучать азам профессии.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [77]:
# текст запроса
n1 = 1
n2 = 5
query_5_1 = f'''(select
                count(v.id),
                e.name
                from vacancies v
                join employers e on v.employer_id = e.id
                group by e.id, e.name
                order by count(v.id) desc
                offset {n1 - 1}
                limit 1)
                union all
                (select
                count(v.id),
                e.name
                from vacancies v
                join employers e on v.employer_id = e.id
                group by e.id, e.name
                order by count(v.id) desc
                offset {n2 - 1}
                limit 1)
'''

In [78]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

/tmp/ipykernel_11359/3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


count           name
0   1933         Яндекс
1    331  Газпром нефть

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [79]:
# текст запроса
query_5_2 = f'''(select
                a.name,
                count(v.id),
                count(e.id)
                from areas a
                left join employers e on e.area = a.id
                left join vacancies v on v.area_id = a.id
                where v.id is null
                group by a.name
                order by 3 desc
                limit 1)
'''

In [80]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

/tmp/ipykernel_11359/2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  count  count
0  Россия      0    410

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [81]:
# текст запроса
query_5_3 = f'''select
                e.name,
                count(distinct v.area_id) cnt
                from vacancies v 
                join employers e on e.id = v.employer_id
                group by e.name
                order by cnt desc
'''

In [82]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

/tmp/ipykernel_11359/201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  cnt
0                      Яндекс  181
1                  Ростелеком  152
2                  Спецремонт  116
3      Поляков Денис Иванович   88
4                   ООО ЕФИН    71
...                       ...  ...
14761                  UniSol    1
14762            UNISTORY LLC    1
14763                   UNIT6    1
14764     United Distribution    1
14765              UNITEL LLC    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [83]:
# текст запроса
query_5_4 = f'''select
                count(e.id)
                from employers e
                left join employers_industries ei on e.id = ei.employer_id
                where ei.industry_id is null
'''

In [84]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

/tmp/ipykernel_11359/3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [85]:
# текст запроса
query_5_5 = f'''select
                e.name
                from employers e
                join employers_industries ei on e.id = ei.employer_id
                group by e.name
                having count(ei.industry_id) = 4
                order by e.name
                offset 2
                limit 1
'''

In [86]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

/tmp/ipykernel_11359/388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [87]:
# текст запроса
query_5_6 = f'''select
                count(distinct ei.employer_id)
                from employers_industries ei
                join industries i on ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения'
'''

In [88]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

/tmp/ipykernel_11359/821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

6.1 Дополнительно. Составим рейтинг сфер деятельности.
 

In [89]:
#текст запроса
query_5_6_1 = f'''select
                count(distinct ei.employer_id) cnt_emp,
                i.name
                from employers_industries ei
                join industries i on ei.industry_id = i.id
                group by i.name
                order by cnt_emp desc
'''                

In [90]:
#результат запроса
df = pd.read_sql_query(query_5_6_1, connection)
df

/tmp/ipykernel_11359/144084129.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6_1, connection)


cnt_emp                                               name
0       3553                Разработка программного обеспечения
1       2993  Системная интеграция,  автоматизации технологи...
2       1675  Интернет-компания (поисковики, платежные систе...
3        798  Маркетинговые, рекламные, BTL, дизайнерские, E...
4        662                              Консалтинговые услуги
..       ...                                                ...
289        4                      Благотворительная организация
290        4        Лифтовое хозяйство (монтаж, сервис, ремонт)
291        3              Ботанический сад, зоопарк, заповедник
292        3                                          Автошкола
293        1             Ассоциация в сфере культуры, искусства

[294 rows x 2 columns]

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [91]:
# код для получения списка городов-милионников

In [92]:
# текст запроса

In [93]:
# результат запроса

***

# Выводы по анализу работодателей
На первом месте по количеству размещенных вакансий (1933) находится компания Яндекс. Причем, отрыв от второго места (Ростелеком) почти в 4 раза (у Ростелекома 491 вакансия). Яндекс предлагает работу в 181 городе (регионе), но всего одной сфере деятельности (интернет-компания). Впрочем, все свои виды деятельности Яндекс включил в это наименование. 
В пятерку лидеров по количеству вакансий, кроме Яндекса и Ростелекома, входят Тинькофф, Сбер и Газпром нефть. Все компании крупные и быстроразвивающиеся. Этим и объясняется их лидерство в количестве размещенных вакансий.
Больше трети работодателей (8419) не указыают свою сферу деятельности, чем усложняют задачу поиска их вакансий для соискателей. 410 работодателей указали в качестве своего родного региона "Россия". Скорее всего, это компании, работающие по всей стране или через интернет, не разобрались, как правильно зарегистрировать свою компанию на HH. Или же это зарубежные компании, ищущие сотрудников в России.
Самой многочисленной сферой деятельности среди работодателей является "Разработка программного обеспечения" (3553 работодателя из 23501), на втором месте "Системная интеграция, автоматизация технологических процессов, ИТ-консалтинг", на третьем - "Интернет-компания". То есть тройку лидеров по количеству вакансий занимают компании, связанные с IT.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [94]:
# текст запроса
query_6_1 = f'''select
                count(id) cnt_vac
                from vacancies
                where name ilike '%data%' or name ilike '%данн%'
''' 

In [95]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

/tmp/ipykernel_11359/366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


cnt_vac
0     1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [96]:
# Дополнительно. Вычислим, сколько вакансий есть для дата-сайентиста всего.
# текст запроса
ds = '''(name ilike '%data scientist%'
        or name ilike '%data science%'
        or name ilike '%исследователь данных%'
        or name ilike '%machine learning%'
        or name ilike '%машинн%обучен%'
        or (name like '%ML%'
            and name not like '%HTML%')
)'''

query_6_2_1 = f'''select
                count(id)
                from vacancies
                where {ds}
''' 


In [97]:
# результат запроса
df = pd.read_sql_query(query_6_2_1, connection)
df

/tmp/ipykernel_11359/4273938854.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2_1, connection)


count
0    480

In [98]:
# теперь вычислим, сколько вакансий для Junior-дата-сайентистов
# текст запроса
query_6_2 = f'''select
                count(id)
                from vacancies
                where {ds}
                and (
                    name ilike '%junior%'
                    or experience = 'Нет опыта'
                    or employment = 'Стажировка'
                    )
''' 

In [99]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

/tmp/ipykernel_11359/2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [100]:
# текст запроса
query_6_3 = f'''select
                count(id)
                from vacancies
                where {ds}
                and (
                key_skills like '%SQL%'
                or key_skills like '%postgres%')
            
''' 

In [101]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

/tmp/ipykernel_11359/3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS. Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [102]:
# текст запроса
query_6_4 = f'''select
                count(id)
                from vacancies
                where {ds}
                and key_skills ilike '%Python%'
            
''' 

In [103]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

/tmp/ipykernel_11359/4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [104]:
# текст запроса
# Вычислим разницу между количеством символов в строке с навыками и количеством символов
# в той же строке, только с удаленными знаками табуляции. Получим количество табуляций. Прибавив 1,
# получим количество слов в строке.
query_6_5 = f'''select
                round(avg((length(key_skills) - length(replace(key_skills, chr(9), '')) +1)), 2)
                from vacancies
                where {ds}
            
''' 

In [105]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

/tmp/ipykernel_11359/3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [106]:
# текст запроса
query_6_6 = f'''select
                experience,
                round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from)) / 2)) salary
                from vacancies
                where {ds}
                     and(
                       salary_to is not null
                       or salary_from is not null
                       )
                group by experience
            
''' 

In [107]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

/tmp/ipykernel_11359/519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience    salary
0           Нет опыта   74643.0
1  От 1 года до 3 лет  139675.0
2       От 3 до 6 лет  243115.0

***

# Выводы по предметному анализу
Вакансии, имеющие отношение к данным, составляют 3,6% от общего количества вакансий на HH. При этом, вакансии для дата-сайентистов составляют чуть больше четверти от этого числа (480 из 1771). А джуниоров-дата-сайентистов ищут в 51 вакансии - это 10,6 % от всех вакансий для DS. 
Новички в DS могут рассчитывать на среднюю заработную плату в размере 74 тыс. рублей. 

# Общий вывод по проекту

In [108]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

# Дополнительно. Вычислим среднюю заработную плату в разрезе требуемого опыта в целом для всех вакансий 
# текст запроса
query_6_6_1 = f'''select
                experience,
                round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from)) / 2)) salary
                from vacancies
                where
                    salary_to is not null
                    or salary_from is not null
                group by experience
                order by salary
            
''' 

In [109]:
# результат запроса
df = pd.read_sql_query(query_6_6_1, connection)
df

/tmp/ipykernel_11359/2364063680.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6_1, connection)


experience    salary
0           Нет опыта   40472.0
1  От 1 года до 3 лет   76541.0
2       От 3 до 6 лет  133211.0
3         Более 6 лет  166256.0

Как видим, начинающий дата-сайентист может рассчитывать на заработную плату, почти в два раза превышающую среднюю зарплату кандидата без опыта среди всех вакансий.

Итак, наша база данных содержит 49197 тыс. вакансий от 23501 работодателей. Наибольшее количество вакансий предлагается в городах-милионниках России и ближнего зарубежья. Средняя зарплатная вилка по вакансиям составляет от 71065 до 110537 рублей. В базе присутствует безусловный лидер по количеству вакансий и регионов их размещения - это компания Яндекс. В целом, ведущие позиции по количеству размещенных вакансий занимают сферы деятельности, связанные с IT. Это означает, что IT-специалисты являются востребованными на рынке труда.

In [110]:
connection.close()